File: zigzag_conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s

        rows = [""] * numRows
        current_row, going_down = 0, False

        for c in s:
            rows[current_row] += c
            if current_row in (0, numRows - 1):
                going_down = not going_down
            current_row += 1 if going_down else -1
        return "".join(rows[i] for i in range(1, numRows - 1))

File: zigzag_conversion.rs
fn convert(s: String, num_rows: usize) -> String {
    if num_rows == 1 {
        return s;
    }
    let mut rows: Vec<String> = vec![String::new(); num_rows as usize];
    let mut current_row = 0;
    let mut going_down = false;
    for c in s.chars() {
        rows[current_row].push(c);
        if current_row == 0 || current_row == num_rows - 1 {
            going_down = !going_down;
        }
        current_row = if going_down {
            current_row + 1
        } else {
            current_row - 1
        };
    }
    rows.into_iter().collect()
}

fn main() {
    let s = "PAYPALISHIRING".to_string();
    let num_rows = 3;
    let result = convert(s, num_rows);
    println!("{}", result);
}
